### ISSUE TITLE
Not loading font for word "SĄD"

### ISSUE THREAD

**Issue by MatiasMovie** [libass/libass#459]
Not loading font for word "SĄD" (first second)
MPV
https://i.imgur.com/HrEUE0x.png
MPC (XySubFilter)
https://i.imgur.com/7m8LsJc.png

**Comment by astiob** [comment]
We can’t help without seeing the MKV or at least the font file.

Your subtitles work for me with a copy of the same font from another file.

**Comment by MatiasMovie** [comment]
Fonts
https://mega.nz/folder/1Qc0TA7A#s97DxtK0UP7b0J-iegNmMw

mkv name
[Die Hot 4 - 4 keepz] Great Pretender - 04 [475596EC].mkv

**Comment by TheOneric** [comment]
On Tue, Dec 08, 2020 at 01:56:54 -0800, MatiasMovie wrote:
> Fonts
> https://mega.nz/folder/1Qc0TA7A#s97DxtK0UP7b0J-iegNmMw
> 
> mkv name
> [Die Hot 4 - 4 keepz] Great Pretender - 04 [475596EC].mkv

Going by the filename’s the linked font folder does _not_ have the required 
“FOT Manyo Sosho Std Stripped EB” font, only a similarly named 
“ManyoGyoshoStd” font (filename: FOTMANYOGYOSHOSTDSTRP-E.otf).

The name of the mkv is irrelevant, the important part is its contents, in this 
case especially its attached fonts. I suspect the mkv is missing the required 
font as well.
You can use eg MKVToolnix to either check the attached fonts and upload 
them if required, or if your not sure what to look for just make a cut (also 
MKVToolnix) of your mkv file and upload eg only the first four seconds of it. 

In case this is is not just a missing font afterall: What system are you using 
libass on (*Linux, MacOS, Windows, …) and if you know: which font provider is 
used (fontconfig, Coretext or DirectWrite)?

**Comment by astiob** [comment]
Neither of those fonts is in the MKV:

```
$ mkvmerge -i "[Die Hot 4 - 4 keepz] Great Pretender - 04 [475596EC].mkv"
File '/home/common/unprotected/transmission/[Die Hot 4 - 4 keepz] Great Pretender - 04 [475596EC].mkv': container: Matroska
Track ID 0: video (MPEG-4p10/AVC/h.264)
Track ID 1: audio (AAC)
Track ID 2: subtitles (SubStationAlpha)
Attachment ID 1: type 'application/x-truetype-font', size 128368 bytes, file name 'Aller Bold.ttf'
Attachment ID 2: type 'application/x-truetype-font', size 35804 bytes, file name 'Bolide Regular.ttf'
Attachment ID 3: type 'application/x-truetype-font', size 34248 bytes, file name 'CNN.TTF'
Attachment ID 4: type 'application/x-truetype-font', size 658076 bytes, file name 'DK Crayon Crumble.ttf'
Attachment ID 5: type 'application/x-truetype-font', size 24984 bytes, file name 'FontinSans-Bold.otf'
Attachment ID 6: type 'application/x-truetype-font', size 24268 bytes, file name 'FontinSans-BoldItalic.otf'
Attachment ID 7: type 'application/x-truetype-font', size 221328 bytes, file name 'OpenSans-Semibold.ttf'
Chapters: 5 entries
Global tags: 5 entries
Tags for track ID 1: 2 entries
```

Are you sure your MPV and XySubFilter screenshots are taken on the same system? It seems most likely that the system where the XySubFilter screenshot was taken has the font installed system-wide, while the MPV screenshot was taken on another system that lacks this font.

**Comment by MatiasMovie** [comment]
Yes, same system, you must install fonts from mega.nz link, becouse mkv not have require fonts

**Comment by MatiasMovie** [comment]
Maybe it's problem with implementation

VLC
https://i.imgur.com/wnoWkUW.png

Kainote (libass render)
https://i.imgur.com/NihNMLA.png

**Comment by TheOneric** [comment]
The file providing “FOT Manyo Sosho Std Stripped EB” in this fontfolder is `Chocolate Covered Raindrops Bold.ttf`, with this in `~/.fonts` and fontconfig as fontprovider I get this:
```
[ass] fontselect: (FOT Manyo Sosho Std Stripped EB, 400, 0) -> ~/.fonts/GREAT PRETENDER/Chocolate Covered Raindrops Bold.ttf, 0, FOTManyoSoshoStdStripped-EB
[ass] Glyph 0x104 not found, selecting one more font for (FOT Manyo Sosho Std Stripped EB, 400, 0)
[ass] fontselect: Using default font family: (FOT Manyo Sosho Std Stripped EB, 400, 0) -> /usr/share/fonts/truetype/croscore/Arimo-Regular.ttf, 0, Arimo
[ass] fontselect: (FOT Manyo Sosho Std Stripped EB, 400, 0) -> /usr/share/fonts/truetype/croscore/Arimo-Regular.ttf, 0, Arimo
```
![grafik](https://user-images.githubusercontent.com/1668471/101522953-399f2a80-3988-11eb-85b2-9f113e329db1.png)

Note that, the font is missing the `Ą`, which is neither present in Manyo Soscho nor in Manyo Gyosho. I don't know where VSF and Kainote are pulling this from. Perhaps an "smart" decomposition of 0x104 into `A` and an combining ogonek? Though Manyo Sosho does _not_ provide a combining nor a non-combining ogonek (0x328, 0x2db) so even then this would need to be supplied by a different font ...

**Comment by astiob** [comment]
Note that the OP’s libass screenshots don’t display the S and the D in the right font, either. The entire font is missing. It all looks as if the Kainote, Aegisub and XySubFilter screenshots are taken on a different system from the VLC and MPV screenshots or with a different copy of the ASS file that embeds the font.

**Comment by astiob** [comment]
I’m not sure how else we can help with this at the moment. It seems the next step towards progress is to identify the exact font file that the Ą actually comes from, and how exactly it is installed on the system. Then we can try to figure out why Aegisub and Kainote are picking it up but VLC and mpv aren’t.

**Comment by MatiasMovie** [comment]
I found I have other font version with similar name and have "Ą"
https://krakenfiles.com/view/ca5074322e/file.html
Maybe that's why XY using this font?

And I found FOT Manyo Sosho Std Stripped EB not install properly (I reinstall) and now looks like this
https://i.imgur.com/CURLjHc.png
but still I don't now why MPC, Kainote and Aegisub using PL version.

**Comment by astiob** [comment]
I can confirm the following things:
* The EB PL font has the Ą.
* VSFilter (multiple flavours) uses the EB PL font if it is attached to an MKV _or_ if it is installed system-wide, regardless of whether the non-PL font is also attached and regardless of which order they’re attached in (installing the non-PL font not tested).
* mpv fails to use the EB PL font in _any_ situation, with Fontconfig and DirectWrite _alike._
* VSFilter produces an interesting result if the non-PL font is attached to an MKV while the PL one isn’t:
![S(space)D](https://user-images.githubusercontent.com/515193/101549979-4424e880-39b7-11eb-8f82-92cbc5b44d37.png)

My conclusions so far:
* This is a genuine VSFilter compatibility issue.
* I have no idea why Kainote and Aegisub reportedly use the font. Are you _sure_ they’re using libass in those screenshots?

**Comment by MatiasMovie** [comment]
Kainote settings
https://i.imgur.com/XBt4mlG.png

Aegisub settings
https://i.imgur.com/XBt4mlG.png

**Comment by bjakja** [comment]
Kainote and Aegisub do not have written `CheckGlyph` method (it returns true) to not show "Arial" font when glyphs are missing. That's why loading fonts is different then in MPV.

**Comment by astiob** [comment]
I’m sorry, but I don’t understand what you’re talking about. There’s no `CheckGlyph` method in either VSFilter or libass, and neither VSFilter nor libass exposes any hooks for the consumer to customize font selection other than adding memory fonts and, for libass, selecting a default font and choosing between Fontconfig or DirectWrite. It doesn’t work in mpv with either Fontconfig or DirectWrite, and I doubt Kainote/Aegisub configure FOT Manyo Sosho Std Stripped EB PL as the default font.

**Comment by astiob** [comment]
OK, I see they actually have _their own font provider_ that they replace libass’s DirectWrite provider with. That’s pretty evil, and explains this discrepancy (and highlights why it’s evil).

**Comment by astiob** [comment]
Now that we know why Aegisub and Kainote use the right font (thanks!), the next step is to figure out why GDI is selecting it in VSFilter, Aegisub and Kainote alike. I have two guesses:

1. Perhaps “EB” is a built-in abbreviation for “Extrabold”, which is the font’s `style`, and GDI is splitting “FOT Manyo Sosho Std Stripped EB” into “FOT Manyo Sosho Std Stripped” and “EB” and matching the parts to `family` and `style` respectively.

2. Perhaps GDI is trying to find the font with the last word dropped for whatever reason.

I seem to remember libass used to iteratively drop trailing words from the font name when searching for the font in an attempt to split the name into family and style, but this must have been before our new fontselect mechanism as I haven’t been able to find this in the current code or its history. Normally I’d assume we stopped doing this because it was incompatible with GDI, but maybe it _was,_ after all?

---

Tangential note re: Aegisub and Kainote’s GDI font provider: I’d love to see a GDI font provider in libass proper (and, frankly, it’s a little disappointing to see Aegisub has had one for five years but no attempt has been made to submit it upstream, unless I forget one). I’d wanted a GDI provider from the beginning, but IIRC I was told making one would be too hard and DirectWrite was easier. Admittedly, Aegisub’s provider lacks some features, but it doesn’t seem complicated, so maybe it would be possible to complete it and add it to libass, possibly even replacing the DirectWrite one?

Also, I see that font provider justifies a lack of fallback by saying it helps authors specify their fonts exactly. In that same spirit, perhaps Kainote (and Aegisub) would be interested in making the font provider even more strict and match only exact full/PostScript names as appropriate. And even better on top of that, somehow ignore system fonts and only allow fonts attached to the ASS or the MKV, to try to prevent subtitle releases using unattached fonts. I realize the programs don’t generally work with whole MKVs, so I’m not sure how exactly this could be accomplished. A nuclear option is to require fonts to be embedded into the ASS file. Perhaps an explicit font whitelist could be a more moderate solution.

**Comment by bjakja** [comment]
Fonts have 3 names, and when only one is changed to PL than windows can return font with PL in name as without PL.

**Comment by astiob** [comment]
None of this font’s names is “FOT Manyo Sosho Std Stripped EB”.

**Comment by TheOneric** [comment]
On Wed, Dec 09, 2020 at 07:36:17 -0800, Oleg Oshmyan wrote:
> None of this font’s names is “FOT Manyo Sosho Std Stripped EB”.

Just speculating, but the warnings FontForge gives about this font may be 
relevant to GDI’s font selection.

First, its postscript name “FOTManyoSoshoStdStripped-EB PL” is apparently 
invalid as spaces are not allowed. If Windows were to abort reading the PS 
name on first invalid character this would yield “FOTManyoSoshoStdStripped-EB” 
without the “PL”.

Second, on editing the base name, I'm warned that “some versions of Windows 
may refuse to install fonts if the familyname is longer than 31 characters”.
The family name being “FOT Manyo Sosho Std Stripped EB PL”, truncating this to 
31 characters again just happens to get rid of the “PL” part.

**Comment by astiob** [comment]
Ooh, that last one sounds most plausible to me. Nice find. I’ll run a few tests.

**Comment by astiob** [comment]
Before I even got to any tests…

```c
typedef struct tagLOGFONTW {
  ...
  WCHAR lfFaceName[LF_FACESIZE];
} LOGFONTW, *PLOGFONTW, *NPLOGFONTW, *LPLOGFONTW;
```

`LF_FACESIZE` is 32.

> A null-terminated string that specifies the typeface name of the font. The length of this string must not exceed 32 **TCHAR** values, including the terminating **NULL**.

([source](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-logfontw))

VSFilter:

```c
strcpy_s(lfa.lfFaceName, CStringA(s.fontName));
```

Aegisub/Kainote:

```c++
MultiByteToWideChar(CP_UTF8, 0, name, -1, lf.lfFaceName, LF_FACESIZE);
```

We have our answer. Thanks! and congratulations!

**Comment by astiob** [comment]
(So the problem is not necessarily with GDI itself, but with the way VSFilter, Aegisub and Kainote invoke it. I don’t know whether there’s a way to supply a longer name to GDI.)

Now, solving this in libass poses a bit of a challenge. It should be easy for font providers that gather the full list of fonts upfront: just truncate the name when going through the list during matching. But at least one (CoreText) queries fonts from the operating system lazily, passing in the desired name specified in the ASS. Now, if the ASS name is at least 31 characters long, we want to find all system fonts whose names start with the same 31 characters…

**Comment by rcombs** [comment]
> I don’t know whether there’s a way to supply a longer name to GDI.

Nope, the internal struct also has a name field with a size of `LF_FACESIZE`, and that's what it matches against.

> But at least one (CoreText) queries fonts from the operating system lazily, passing in the desired name specified in the ASS.

Honestly I'm not that concerned for this. If the font isn't muxed, I don't think it's super essential that we always match bug-compatibly with GDI. Once we're querying system-installed fonts, all bets are off wrt compatibility (since there's no guarantee the user has the same font installed, or that an installed font with a matching name will be the exact one the script author used), and that's going to be true no matter what we do here. IMO, system-font fallback is on a best-effort basis, with the only guarantee being that we'll find _something_ to render a glyph with if possible.

**Comment by astiob** [comment]
For reference, our CoreText font provider used to eagerly query all fonts in its initial implementation, but this was changed to lazy loading in b0776d353f8c3396165effd21e113c4f225daa0d, with the old code behind a macro toggle until it was removed in 77fde472621c6ef263f9b5a7678e3a8677a2eb46. The original commit message says lazy loading is faster and supports this claim with the following numbers:

    CT_FONTS_EAGER_LOAD=0
    0.04s user 0.02s system 79% cpu 0.081 total
    
    CT_FONTS_EAGER_LOAD=1
    0.12s user 0.06s system 44% cpu 0.420 total

I assume this means an extra 80 ms delay on startup, which doesn’t seem terribly slow to me, although being even faster is, of course, even better.

**Comment by astiob** [comment]
I’ve updated my [earlier comment](https://github.com/libass/libass/issues/459#issuecomment-741867672) with further corrections:

* Modern VSFilters, indeed, truncate font names to 31 characters.
* Ancient VSFilter did not: it simply didn’t _support_ names longer than 31 characters.
* Aegisub/Kainote’s GDI font provider for libass also _doesn’t support_ names longer than 31 characters (and does not truncate them).

**Comment by astiob** [comment]
Corrected the correction:

A simple test shows that `MultiByteToWideChar` does, in fact, truncate the UTF-16 output string (even in the middle of a surrogate pair), even as it reports an error and 0 as the return value. The reason longer names don’t work with Aegisub/Kainote is that the font returned by GDI has its name truncated, so the font provider returns the truncated name to libass’s fontselect, which then fails to find a match.
